specification/dartLangSpec.tex

```diff
@@ -10372,12 +10374,50 @@ \subsubsection{Binding Actuals to Formals}
 It is a compile-time error if $q_j = q_k$ for any $j \ne k$.
 
 \LMHash{}%
-If the static type of $f$ is \DYNAMIC{} or the built-in class \FUNCTION{},
-no further static checks are performed and the static type of $i$ is \DYNAMIC{};
-otherwise, it is a compile-time error if the static type of $f$ is not a function type.
+For a given type $T_0$, we introduce the notion of a
+\IndexCustom{$T_0$ bounded type}{type!T0 bounded}:
+$T_0$ itself is $T_0$ bounded;
+if $B$ is $T_0$ bounded and
+$X$ is a type variable with bound $B$
+then $X$ is $T_0$ bounded;
+finally, if $B$ is $T_0$ bounded and
+$X$ is a type variable
+then $X \& B$ is $T_0$ bounded.
+In particular, a
+\IndexCustom{\DYNAMIC{} bounded type}{type!dynamic bounded}
+is either \DYNAMIC{} itself
+or a type variable whose bound is \DYNAMIC{} bounded,
+or an intersection whose second operand is \DYNAMIC{} bounded.
+Similarly for a
+\IndexCustom{\FUNCTION{} bounded type}{type!function bounded}.
+
+\LMHash{}%
+Similarly, we introduce the notion of a
+\IndexCustom{function-type bounded type}{type!function-type bounded}:
+Every function type $T$
+(\ref{functionTypes})
+is a function-type bounded type,
+with associated function type $T$;
+if $B$ is function-type bounded with associated function type $T$ and
+$X$ is a type variable with bound $B$
+then $X$ is function-type bounded with associated function type $T$;
+finally, if $B$ is function-type bounded with associated function type $T$ and
+$X$ is a type variable
+then $X \& B$ is function-type bounded with associated function type $T$.
+
+\LMHash{}%
+If the static type of $f$ is \DYNAMIC{} bounded or \FUNCTION{} bounded,
+no further static checks are performed on the invocation $i$
+(\commentary{apart from separate static checks on subterms like arguments}),
+and the static type of $i$ is \DYNAMIC{}.
+Otherwise, it is a compile-time error if the static type of $f$ is not
+function-type bounded.
+
+\LMHash{}%
+Otherwise, the static type of $f$ is function-type bounded
```

Comment:
Feels redundant with the "Otherwise" above. Maybe just grammatically, but I find it hard to understand.

Consider rewriting the above to;

> It's a compile-time error if the static type of $f$ is not a `dynamic`-bounded, `Function`-bounded 
> or function-type bounded type.

Maybe start the paragraph with that, then handle `dynamic` and `Function`, then this "Otherwise" fits right into the flow.

That avoids the two "Otherwise"s in a row.

Reply:
Rewritten to avoid the awkward 'Otherwise'.